### DISCUSSION TITLE
Could not find data file data/data/abbreviations with `pandoc server`

### DISCUSSION THREAD

**Discussion by larstvei** [jgm/pandoc#10024] [Q&A]
Hi!

I am unable to run the `pandoc server` successfully. I run it with:

```sh
λ pandoc server
Starting server on port 3030...
```

If I try to interact with it, I get the following error:

```sh
λ curl -X POST http://localhost:3030 \
           -H "Content-Type: application/json" \
           -d '{
             "text": "# Hello Pandoc",
             "from": "markdown",
             "to": "html"
           }'
Could not find data file data/data/abbreviations
```

I am running Pandoc installed via nix on macOS. The problem is not macOS-related, as I have reproduced the issue on Linux as well (with two different versions of Pandoc).

Here is the output of my version:

```sh
λ pandoc -v
pandoc 3.1.11.1
Features: +server +lua
Scripting engine: Lua 5.4
User data directory: /Users/larstvei/.local/share/pandoc
Copyright (C) 2006-2023 John MacFarlane. Web: https://pandoc.org
This is free software; see the source for copying conditions. There is no
warranty, not even for merchantability or fitness for a particular purpose.
```

The user data directory is _empty_, which is almost certainly the issue. However, I think this is intentional, as Pandoc is configured to be built with [data files embedded](https://github.com/NixOS/nixpkgs/blob/24899eea991218f4cfa678024b0821268b9d9670/pkgs/development/tools/pandoc/default.nix#L9).

Has anyone gotten `pandoc server` to run with nix? Or is someone else able to reproduce this issue?

I think this is a bug, but I am not entirely sure if it should be filed as at pandoc or nixpkgs.

**Comment by jgm** [comment]
The problem does not occur with the pandoc installed using our package installer.  This leads me to think it's a nix package issue, but I'm not sure what it would be...

**Comment by one-d-wide** [comment]
I compiled pandoc from source with -f-embed_data_files, and it has the same issue.

```sh
$ cabal install pandoc-cli -f-embed_data_files

$ echo "``test``" | ~/.cabal/bin/pandoc -f rst -t markdown -o - -
`test`

$ ~/.cabal/bin/pandoc server &
Starting server on port 3030...
$ curl -X POST --json '{"text": "``test``", "from": "rst", "to": "markdown"}' -o - http://127.0.0.1:3030
{"error":"Could not find data file data/data/abbreviations"}

$ strings ~/.cabal/bin/pandoc | grep -C5 -- "saec\." || echo "No plaintext abbreviations"
No plaintext abbreviations
```

In case it's something weird with NixOS, I checked Gentoo package - same results.

**Reply by jgm** [reply]
In `-f-embed_data_files`, the `-` before the flag name means "UNSET this flag."
Try with `-fembed_data_files` or `-f+embed_data_files`.

**Reply by one-d-wide** [reply]
> In `-f-embed_data_files`, the `-` before the flag name means "UNSET this flag."

Yeah, that what I wanted to show. I know at least 2 packages (NixOS and Gentoo) not enabling this flag, hence `pandoc server` is broken[^1] there. So maybe it could be fixed or clarified (or printed as warning) that pandoc server requires embeddings.

[^1]: Well, it kinda works, but you need to add `"files": {"data/data/abbreviations": ""}` to the request, which is not immediately obvious.

**Reply by jgm** [reply]
It doesn't require embeddings.  However, if it is not compiled with `embed_data_files`, then it needs to have the system data files available at a certain path (configurable using `--datadir` with cabal).  For example, debian packages pandoc with data files and ensures they are put where pandoc needs to find them.  It is certainly easier for packagers to use `embed_data_files`, but if they want to avoid it, they need to use the right flags and move the data files to the required place.

**Reply by one-d-wide** [reply]
Hm, I think cabal is already supplied with --datadir on NixOS  (though I'm in no way expert on cabal builds), as it produces a separate *-data/ output, which breaks `pandoc -t markdown` if removed.

I tried setting --datadir in manual build, but pandoc server still doesn't work:

```sh
$ cabal configure pandoc-cli --datadir=$(pwd)/data -f-embed_data_files
$ cabal build pandoc-cli
[...]
[4 of 4] Linking .../pandoc
$ echo "``test``" | .../pandoc -f rst -t markdown -o - -
Could not find data file [pwd]/data/x86_64-linux-ghc-9.10.3-a85d/pandoc-3.9.0.2/data/abbreviations
$ mkdir -p data/x86_64-linux-ghc-9.10.3-a85d/pandoc-3.9.0.2
$ ln -s $(pwd)/data data/x86_64-linux-ghc-9.10.3-a85d/pandoc-3.9.0.2/data
$ echo "``test``" | .../pandoc -f rst -t markdown -o - -
`test`
$ .../pandoc server &
Starting server on port 3030...
$ curl -X POST --json '{"text": "test``", "from": "rst", "to": "markdown"}' -o - http://127.0.0.1:3030
{"error":"Could not find data file data/data/abbreviations"}
```

**Reply by jgm** [reply]
You may need to use `cabal install` to get the data dir placed where it is expected?  Consult the cabal manual.  I always embed, so I've lost track of how this works exactly.

**Comment by jgm** [comment]
Regarding the nix issue: note that `embed_data_files` needs to be set not for the compilation of `pandoc-cli`, but for the compilation of `pandoc`.  That may perhaps be the issue?
